### PR TITLE
support cross/outer apply and skip/take in oracle V12c

### PR DESCRIPTION
**Pull request by jogibear9988**

### CODE REVIEW

File: Source/LinqToDB/DataProvider/Oracle/OracleDataProvider.cs [hunk 1]
```diff
@@ -14,27 +14,40 @@ namespace LinqToDB.DataProvider.Oracle
...
+			Version = version;
+
 			//SqlProviderFlags.IsCountSubQuerySupported        = false;
 			SqlProviderFlags.IsIdentityParameterRequired       = true;
 			SqlProviderFlags.IsCommonTableExpressionsSupported = true;
 			SqlProviderFlags.IsSubQueryOrderBySupported        = true;
 			SqlProviderFlags.IsDistinctOrderBySupported        = false;
 			SqlProviderFlags.IsUpdateFromSupported             = false;
 
+			if (version != OracleVersion.v11)
```
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/Oracle/OracleDataProvider.cs]: Replace with
```cs
version >= OracleVersion.v12
```

File: Source/LinqToDB/DataProvider/Oracle/OracleVersion.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace LinqToDB.DataProvider.Oracle
+{
+	public enum OracleVersion
+	{
+		v11,
+		v12,
+		latest
```
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/Oracle/OracleVersion.cs]: replace with `Latest`

**Comment by sdanyliv** [comment]
Still needed Version Detector, as it is done for PostgresSQL and SQL Server
```sql
select VERSION from PRODUCT_COMPONENT_VERSION
where PRODUCT like 'PL/SQL%';
```

**Comment by jogibear9988** [comment]
@MaceWindu 


> Still needed Version Detector, as it is done for PostgresSQL and SQL Server
> 
> ```sql
> select VERSION from PRODUCT_COMPONENT_VERSION
> where PRODUCT like 'PL/SQL%';
> ```

I've implemented something, but untested at the moment. Need to look when I find time to test!

File: Source/LinqToDB/DataProvider/Oracle/OracleDataProvider.cs [hunk 3]
```diff
@@ -14,27 +14,40 @@ namespace LinqToDB.DataProvider.Oracle
...
 			SqlProviderFlags.MaxInListValuesCount              = 1000;
 
 			SetCharField            ("Char",  (r, i) => r.GetString(i).TrimEnd(' '));
 			SetCharField            ("NChar", (r, i) => r.GetString(i).TrimEnd(' '));
 			SetCharFieldToType<char>("Char",  (r, i) => DataTools.GetChar(r, i));
 			SetCharFieldToType<char>("NChar", (r, i) => DataTools.GetChar(r, i));
 
-			_sqlOptimizer = new OracleSqlOptimizer(SqlProviderFlags);
+			if (version == OracleVersion.v11)
+				_sqlOptimizer = new OracleSqlOptimizer(SqlProviderFlags);
```
Comment (MaceWindu) [inline Source/LinqToDB/DataProvider/Oracle/OracleDataProvider.cs]: We should use 11 in provider classes for v11

File: Source/LinqToDB/DataProvider/Oracle/OracleVersion.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace LinqToDB.DataProvider.Oracle
+{
+	public enum OracleVersion
+	{
+		v11 = 11,
+		v12 = 12,
+		v18 = 18,
```
Comment (MaceWindu) [inline Source/LinqToDB/DataProvider/Oracle/OracleVersion.cs]: We shouldn't add versions, we don't use

File: Source/LinqToDB/DataProvider/Oracle/OracleDataProvider.cs (hunk 3 above)
Comment (jogibear9988) [reply Source/LinqToDB/DataProvider/Oracle/OracleDataProvider.cs]: Don't know what you mean here...

Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/Oracle/OracleDataProvider.cs]: OracleSqlOptimizer -> Oracle11SqlOptimizer

File: Source/LinqToDB/DataProvider/Oracle/Oracle11SqlBuilder.cs [hunk 5]
```diff
@@ -97,6 +97,12 @@ protected override void BuildPredicate(ISqlPredicate predicate)
 			base.BuildPredicate(predicate);
 		}
 
+		protected bool BasiSqlBuilderBuildWhere(SelectQuery selectQuery)
```
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/Oracle/Oracle11SqlBuilder.cs]: Why it is needed?

Comment (jogibear9988) [reply Source/LinqToDB/DataProvider/Oracle/Oracle11SqlBuilder.cs]: So I can call the Base Function from V12 provider. So if smth is added to base V12 also has included this.

**Comment by jogibear9988** [comment]
@MaceWindu @sdanyliv 
I think I'm finished with this feature.
should I rebase the commits into one? 
And is something left I should do?

**Comment by sdanyliv** [comment]
@jogibear9988, i see regression at least in one test
TruncateIdentityNoResetTest

**Comment by jogibear9988** [comment]
I don't understand what I should have changed so this test will fail. Cause the Code for oracle 11 should be the same.
And the test works on MacOs with V11... strange..

**Comment by MaceWindu** [comment]
TruncateIdentityNoResetTest IIRC is unstable test, which fails from time to time waiting somebody to lok at it :)

**Comment by jogibear9988** [comment]
maybe you can review my code. for me there is no change wich can cause it to fail.

**Comment by sdanyliv** [comment]
Doing that right now. Failing test is really strange and contain long standing bug.

File: Source/LinqToDB/DataProvider/Oracle/Oracle11SqlBuilder.cs (hunk 5 above)
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/Oracle/Oracle11SqlBuilder.cs]: Remove that it do nothing.

Comment (jogibear9988) [reply Source/LinqToDB/DataProvider/Oracle/Oracle11SqlBuilder.cs]: I only put it there to use it in v12 provider to access the method from the base class. But I will remove it.

**Pull request merged**